### PR TITLE
Dockerize Community App

### LINKED ISSUE #3106: Dockerize Community Web App
Build docker image for Web App to ease deployment

### PR DESCRIPTION
**Pull request by awasum**
Dockerize Community App.
Update readme with intructions on how to build.
Docker image is created and app is run in nginx when image is executed.

Fix #3106

### CODE REVIEW

**Comment by awasum** [comment]
I wonder why this is failing. Does anyone know why?

File: README.md [hunk 1]
```diff
@@ -119,6 +119,21 @@ Start a static server and open the project in the default browser. The applicati
...
+To build a Docker image for the current repo, run:
+```
+docker build -t mifos-community-app .
+```
+You can then run a Docker Container from the image above like this:
+```
+docker run --name mifos-ui -it -d -p 80:80 mifos-community-app
+```
+
+Access the webapp on http://localhost in your browser. Ma
```
Comment (vorburger) [inline README.md:133]:
```suggestion
Access the webapp on http://localhost in your browser.
```

File: README.md [hunk 2]
```diff
@@ -119,6 +119,21 @@ Start a static server and open the project in the default browser. The applicati
...
+The Dockerfile uses a ruby and node base image to build the current repo and deploys the app on nginx which is exposed 
+port 80 within the container
```
Comment (vorburger) [inline README.md:135]:
```suggestion
on port 80 within the container.
```

**Review by vorburger — COMMENTED** [review]
> I wonder why this is failing. Does anyone know why?

Isn't [the Travis build failure](https://travis-ci.com/openMF/community-app/builds/117581158) completely unrelated to this PR? This PR just adds a Dockerfile. That failure is Travis broken - without Docker, just "plain" Travis (which does something similar to what you're adding here - BTW perhaps it would be possible to share the steps?). I just raised #3110, and expect it to fail similarly.

**Comment by vorburger** [comment]
@awasum may I make a suggestion? I noticed how the steps you're adding into the Dockerfile and what the README describes for manual steps and [what's in the .travis.yml](https://github.com/openMF/community-app/blob/develop/.travis.yml) are, naturally, very similar... how about extracting that from `.travis.yml` out into a (new) `build.sh` and making `.travis.yml` just call that `build.sh` (in a new PR, separate from this one; once the basic Travis build itself is fixed again), and then in your `Dockerfile` introduced in this PR just `COPY` and `RUN` that `build.sh` - do you know what I mean? I feel that could reduce future double maintenance hassles.

File: README.md (hunk 1 above)
Comment (awasum) [reply README.md:133]: Thanks

File: README.md (hunk 2 above)
Comment (awasum) [reply README.md:135]: Thanks

**Comment by awasum** [comment]
> @awasum may I make a suggestion? I noticed how the steps you're adding into the Dockerfile and what the README describes for manual steps and [what's in the .travis.yml](https://github.com/openMF/community-app/blob/develop/.travis.yml) are, naturally, very similar... how about extracting that from `.travis.yml` out into a (new) `build.sh` and making `.travis.yml` just call that `build.sh` (in a new PR, separate from this one; once the basic Travis build itself is fixed again), and then in your `Dockerfile` introduced in this PR just `COPY` and `RUN` that `build.sh` - do you know what I mean? I feel that could reduce future double maintenance hassles.

Yeah, I think I understand you. create the build.sh with similar steps and call the file in travis.yml and in docker for consistency.

**Comment by vorburger** [comment]
It's cool to see that the Dockerfile here is "multi stage", just like just proposed over in [apache/fineract#624](https://github.com/apache/fineract/pull/624) for the back-end, and cleanly separates the builder image running all the fancy JS Node/whatever build related stuff and the runtime image which "just" serves the (static) result.

PS: Duh, I hadn't seen this before; thus preivous (now deleted) comment.

**Review by vorburger — CHANGES_REQUESTED** [review]
@awasum did you mean to apply suggested (minor..) improvements, or were you hoping thisfor this to get merged as is already?

**Comment by awasum** [comment]
> @awasum did you mean to apply suggested (minor..) improvements, or were you hoping thisfor this to get merged as is already?

Will add your improvements before merging....

**Comment by awasum** [comment]
> @awasum I don't mean to "pile on" more (it's super cool that you are working on this!), but in an ideal world, I'd love the Dockerfile here to be "multi stage" (see https://docs.docker.com/develop/develop-images/multistage-build/ ...) as well, just like just proposed over in [apache/fineract#624](https://github.com/apache/fineract/pull/624) for the back-end, and so cleanly separate the builder image running all the fancy JS Node/whatever build related stuff and the runtime image which "just" serves the (static) result.

I am not sure I understand completely. I am using multi stage docker builds. Maybe am wrong cus I have ruby/node as builder to build the product package and then am copying from builder in nginx. Is there a better way? Please advice.. I hope to get this done this night. My time is GMT + 1.

**Comment by awasum** [comment]
> It's cool to see that the Dockerfile here is "multi stage", just like just proposed over in [apache/fineract#624](https://github.com/apache/fineract/pull/624) for the back-end, and cleanly separates the builder image running all the fancy JS Node/whatever build related stuff and the runtime image which "just" serves the (static) result.
> 
> PS: Duh, I hadn't seen this before; thus preivous (now deleted) comment.

Yes...I realised this. Thanks.

**Comment by awasum** [comment]
@vorburger ..I will delay with this till weekend..I have some work which just showed up at day job...Please..

Or if you have the time..go ahead and fix this and merge...

**Comment by diegoc-am** [comment]
Hi, is this going to be updated? I'd like to take care of the changes if it's not

**Comment by awasum** [comment]
Go ahead. @belfazt

**Comment by diegoc-am** [comment]
Corrections can be found here https://github.com/openMF/community-app/pull/3117

**Comment by awasum** [comment]
Do you want to also address this in this in that PR? https://github.com/openMF/community-app/pull/3107#issuecomment-510361027

**Comment by diegoc-am** [comment]
I'd prefer to do that in a separate PR and then just apply that change to the `Dockerfile` and `.travis.yml` once this is merged, which I might do later

**Comment by awasum** [comment]
Ok...Let Travis build.

**Comment by vorburger** [comment]
This was merged in #3117 as dffcb3f1307a8311d17be8ae297c7ce70d3ad009.

I'd love to see a follow-up PR from @belfazt or @awasum re. unifying things; keeping #3106 open for that.

**Pull request closed without merging**